tensorflow_lite_support/cc/test/task/processor/image_preprocessor_test.cc

```diff
@@ -0,0 +1,131 @@
+using ::tflite::task::core::TfLiteEngine;
+using ::tflite::task::vision::DecodeImageFromFile;
+using ::tflite::task::vision::FrameBuffer;
+using ::tflite::task::vision::ImageData;
+
+constexpr char kTestDataDirectory[] =
+    "tensorflow_lite_support/cc/test/testdata/task/vision/";
+
+constexpr char kDilatedConvolutionModelWithMetaData[] = "dilated_conv.tflite";
+
+StatusOr<ImageData> LoadImage(std::string image_name) {
+  return DecodeImageFromFile(
+      JoinPath("./" /*test src dir*/, kTestDataDirectory, image_name));
+}
+
+class DynamicInputTest : public tflite_shims::testing::Test {
+ public:
+  void SetUp() {
+    engine_ = absl::make_unique<TfLiteEngine>();
+    engine_->BuildModelFromFile(JoinPath("./", kTestDataDirectory,
+                                         kDilatedConvolutionModelWithMetaData));
+    engine_->InitInterpreter();
+
+    SUPPORT_ASSERT_OK_AND_ASSIGN(preprocessor_,
+                                 ImagePreprocessor::Create(engine_.get(), {0}));
+
+    SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image, LoadImage("burger.jpg"));
+    frame_buffer_ = CreateFromRgbRawBuffer(
+        image.pixel_data, FrameBuffer::Dimension{image.width, image.height});
+
+    preprocessor_->Preprocess(*frame_buffer_);
+  }
+
+ protected:
+  std::unique_ptr<TfLiteEngine> engine_ = nullptr;
+  std::unique_ptr<FrameBuffer> frame_buffer_ = nullptr;
+  std::unique_ptr<ImagePreprocessor> preprocessor_ = nullptr;
+};
+
+// See if output tensor has been re-dimmed as per the input
+// tensor. Expected shape: (1, input_height, input_width, 16).
+TEST_F(DynamicInputTest, OutputDimensionCheck) {
+  EXPECT_TRUE(engine_->interpreter_wrapper()->InvokeWithoutFallback().ok());
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[0], 1);
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[1],
+            engine_->GetInputs()[0]->dims->data[1]);
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[2],
+            engine_->GetInputs()[0]->dims->data[2]);
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[3], 16);
+}
+
+// Compare pre-processed input with an already pre-processed
+// golden image.
+TEST_F(DynamicInputTest, GoldenImageComparison) {
+  // Get the processed input image.
+  float* processed_input_data =
+      tflite::task::core::AssertAndReturnTypedTensor<float>(
+          engine_->GetInputs()[0]);
+
+  bool is_equal = true;
+
+  const uint8* image_data = frame_buffer_->plane(0).buffer;
+  const size_t input_byte_size =
+      frame_buffer_->plane(0).stride.row_stride_bytes *
+      frame_buffer_->dimension().height;
+
+  for (size_t i = 0; i < input_byte_size / sizeof(uint8);
+       ++i, ++image_data, ++processed_input_data)
+    is_equal &=
+        std::fabs(static_cast<float>(*image_data) - *processed_input_data) <=
+        std::numeric_limits<float>::epsilon();
+
+  EXPECT_TRUE(is_equal);
+}
+
+// Modifying batch/depth to invalid size after Init()
+// call should throw error.
+TEST_F(DynamicInputTest, InvalidBatchOrDepthResize) {
```

Comment:
Again, this won't happen, since ResizeInputTensor() should not be exposed. And we'll most likely only expose const Interpreter from engine.

Reply:
Okay, but bear in mind that `ResizeInputTensorStrict` won't work (which we did in `height_mutable_ || width_mutable_` ) if `const` Interpreter is taken. Perhaps we could do something like this

```c++
// image_preprocessor.cc
engine_->ResizeInputTensorStrict(...)
```

where
```c++
// engine.cc
class TFLiteEngine {
public:
   ResizeInputTensorStrict(..) {
       interpreter_->ResizeInputTensorStrict(...);
}
}
```
That way we can conveniently hide `ResizeInputTensor` and expose `ResizeInputTensorStrict` all the while having `const` Interpreter.

Reply:
Okay, I've changed it to `ResizeInputTensorStrict()` which would be exposed. Tried to resize to invaild batch size, and expected an error.

Reply:
I know it's a little bit confusing, so the point is end users of processors (like your tests here) will only be able to access `engine_wrapper`, not `engine`. But processors can access `engine`, because `engine` is completely internal to Task library. Therefore. You should pretend using `engine_wrapper` for these tests. And this  particular test can be removed, since it won't happen (e.g., you won't be able to  access `ResizeInputTensorStrict` through `engine_wrapper`).

Reply:
Well, that puts things into perspective.